File: Day4Part2.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;
import java.util.stream.Collectors;
import java.util.Map;
import java.util.HashMap;
import java.util.StringTokenizer;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class Day4Part2 {

    public static void main(String[] args) throws Exception {

        long count = new GroupingParser()
            .parse("input-day4.txt")
            .stream()
            .map(Day4Part2::linesToMap)
            .filter(Day4Part2::isValidPassport)
            .count();

        System.out.println(count);
    }

    static boolean isIntegerBetween(int min, int max, String s) {
        try {
            int n = Integer.parseInt(s);
            return min <= n && n <= max;
        }
        catch (Exception e) { return false; }
    }

    static Pattern heightPattern = Pattern.compile("^([0-9]{2,3}+)(in|cm)$");
    static Pattern hairColor = Pattern.compile("^#[0-9a-f]{6}+$");
    static Pattern nineDigits = Pattern.compile("\\d{9}+");
    static List<String> eyeColors = Arrays.asList("amb","blu","brn","gry","grn","hzl","oth");

    static boolean isValidPassport(Map<String,String> passport) {
        if (!isIntegerBetween(1920, 2002, passport.get("byr"))) return false;
        if (!isIntegerBetween(2010, 2020, passport.get("iyr"))) return false;
        if (!isIntegerBetween(2020, 2030, passport.get("eyr"))) return false;
        if (!eyeColors.contains(passport.get("ecl"))) return false;
        if (!hairColor.matcher(passport.getOrDefault("hcl","")).matches()) return false;
        if (!nineDigits.matcher(passport.getOrDefault("pid","")).matches()) return false;

        String height = passport.getOrDefault("hgt","");
        Matcher heightMatcher = heightPattern.matcher(height);
        if (!heightMatcher.matches()) return false;
        String units = heightMatcher.group(1);
        String unitOfMeasure = heightMatcher.group(2);
        if ("cm".equals(unitOfMeasure) && !isIntegerBetween(150,193,units)) return false;
        if ("in".equals(unitOfMeasure) && !isIntegerBetween(59,76,units))   return false;

        return true;
    }

    static Map<String,String> linesToMap(List<String> lines) {
        final Map<String,String> map = new HashMap<>();
        lines.forEach(line -> {
            StringTokenizer st = new StringTokenizer(line);
            while (st.hasMoreElements()) {
                String[] tokens = st.nextToken().split(":");
                map.put(tokens[0], tokens[1]);
            }
        });
        return map;
    }
}

File: day4part2.clj
(defn group [strs]
  (loop [ss strs current [] groups []]
    (let [s (first ss)]
      (if (empty? ss)
        (conj groups current)
        (if (empty? s)
          (recur (rest ss) [] (conj groups current))
          (recur (rest ss) (conj current s) groups))))))

(defn to-map
  [ss]
  (into {}
    (mapv 
      (fn [s] (into [] (.split s ":")))
      (.split (String/join " " ss) " "))))

(defn int-between?
  [x y s]
  (let [n (try (Integer. s)
               (catch Exception e 0))]
    (and (<= x n)
         (>= y n))))

(defn valid-height?
  [s]
  (let [matcher (re-matcher #"([0-9]{2,3}+)(cm|in)" s)]
    (if (.matches matcher)
      (condp = (.group matcher 2)
        "cm" (int-between? 150 193 (.group matcher 1))
        "in" (int-between?  59  76 (.group matcher 1))))))

(defn valid-passport?
  [passport]
  (and
    (int-between? 1920 2002 (get passport "byr"))
    (int-between? 2010 2020 (get passport "iyr"))
    (int-between? 2020 2030 (get passport "eyr"))
    (some #{(get passport "ecl")} ["amb","blu","brn","gry","grn","hzl","oth"])
    (re-matches #"#[0-9a-f]{6}+" (get passport "hcl" ""))
    (re-matches #"\d{9}+" (get passport "pid" ""))
    (valid-height? (get passport "hgt" ""))))
    
(let [passports
      (mapv to-map 
            (group (.split (slurp "input-day4.txt") "\n")))]
  (println (count (filter valid-passport? passports))))


